versions/development/SPEC.md

```diff
@@ -2327,9 +2327,17 @@ Then the command would be instantiated as:
 /bin/mycmd /path/to/c.txt
 ```
 
-## Prepending a String to an Optional Parameter
+In general, it's invalid to supply an expression of optional type `T?` for a non-optional input or variable of type `T`, as the latter cannot accept `None`. (The idiom `select_first([expr, default])` coerces `expr : T?` to `T` by substituting `default : T` when `expr` is undefined; if `default` is omitted, then the coercion fails at runtime.) Similarly, an `Array[T]` expression does not satisfy an `Array[T]+` input.
```

Comment:
A way of coercing `Array[T]` to `Array[T]+` will need to be added.

Reply:
Agreed, it will be problematic to statically enforce `+` without also adding such a coercion. We can also contemplate leaving `+` as more of a / runtime assertion, with the downside of nonuniformly treating `?` and `+`. Discussion on this welcome. cc @orodeh

Reply:
@mlin, thanks for straightening out these issues. 

My personal opinion is that having `+` as a non-empty array qualifier is fine, but I prefer that it's implementation be done as a runtime assertion. If it is part of the type system, then we run into a combinatorial explosion of types. For example:

```
Array[String], Array[String]+
Array[Array[String]], Array[Array[String]]+, Array[Array[String]+], Array[Array[String]+]+
```

However, I would argue that most people find two types sufficient here:  `Array[String]` and `Array[Array[String]]`.

Reply:
I don't mind (/actively like) the fact that you can nest this as a type constraint inside other types. It makes the logic much easier to implement, at least...

I agree though that it should be a coercion (and thus enforced via a runtime check) to go from `Array[X]` to `Array[X]+`. Anything else (eg engine functions) is going to be a nightmare to apply correctly in the more complex cases here

Reply:
@cjllanwarne just clarifying -- do you want to see it enforced as a runtime assertion *rather than* statically? (I'm on the fence at this point)

Reply:
`+` to me has always seemed like a strange type. To me its a more of a runtime decorator that is an assertion check as opposed to a `Type` check.  I for one do not even mind if it is removed from the specification entirely, however that likely would cause people to raise their pitchforks in protest. 

Realistically, the only guarantee that `Array[X]+` provides is that `Array[0]` will not be `None`. This is not a strong enough argument to assert that it is a separate type completely IMO, and it SHOULD be treated as a normal `Array[]` type. Ie the following should be valid with no need to coercion:
```wdl
Array[X]+ myArray
Array[X] otherArray = myArray
```

The one thing we need to be cognisant of however, is that be relegating the `+` type to a `decorator` and not a type, we essentially are introducing a new concept to WDL. That of `Decorator` expressions, which serve as runtime `hints` that should be enforced by the engine, but should not change the typeing of the variable they are on